File: app/schemas/admin_team.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
)

from app.core.constants import AdminRole

# ============================================================
# SHARED
# ============================================================


class AdminTeamSchema(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
    )


# ============================================================
# ADMIN MEMBERS
# ============================================================


class AdminTeamMemberResponse(
    AdminTeamSchema
):
    id: uuid.UUID
    email: str
    display_name: str
    role: AdminRole
    is_active: bool
    mfa_enabled: bool
    last_login_at: datetime | None
    invited_at: datetime | None
    activated_at: datetime | None
    disabled_at: datetime | None
    security_updated_at: datetime | None
    created_at: datetime
    updated_at: datetime


class AdminTeamListResponse(
    AdminTeamSchema
):
    items: list[
        AdminTeamMemberResponse
    ]


# ============================================================
# INVITATIONS
# ============================================================


class CreateAdminInvitationRequest(
    AdminTeamSchema
):
    email: str = Field(
        min_length=3,
        max_length=320,
    )
    role: AdminRole
    expires_at: datetime

    @field_validator(
        "email"
    )
    @classmethod
    def normalize_email(
        cls,
        value: str,
    ) -> str:
        normalized = (
            value.strip().lower()
        )

        if not normalized:
            raise ValueError(
                "Email is required."
            )

        return normalized


class AdminInvitationResponse(
    AdminTeamSchema
):
    id: uuid.UUID
    email: str
    role: AdminRole
    invited_by_admin_id: uuid.UUID
    expires_at: datetime
    created_at: datetime
    updated_at: datetime
    accepted_at: datetime | None
    accepted_admin_id: (
        uuid.UUID | None
    )
    revoked_at: datetime | None
    revoked_by_admin_id: (
        uuid.UUID | None
    )
    revoke_reason: str | None


class AdminInvitationListResponse(
    AdminTeamSchema
):
    items: list[
        AdminInvitationResponse
    ]


class CreatedAdminInvitationResponse(
    AdminTeamSchema
):
    """
    Response returned after an administrator invitation has
    been created and its one-time credential has been sent
    through the configured email provider.

    The raw invitation token must never be returned to the
    administrator browser.
    """

    invitation: AdminInvitationResponse


class RevokeAdminInvitationRequest(
    AdminTeamSchema
):
    reason: str | None = Field(
        default=None,
        max_length=255,
    )


# ============================================================
# INVITATION ACCEPTANCE
# ============================================================


class AcceptAdminInvitationRequest(
    AdminTeamSchema
):
    invitation_token: str = Field(
        min_length=32,
        max_length=1024,
    )

    @field_validator(
        "invitation_token"
    )
    @classmethod
    def normalize_token(
        cls,
        value: str,
    ) -> str:
        normalized = value.strip()

        if not normalized:
            raise ValueError(
                "Invitation token is required."
            )

        return normalized


class AcceptedAdminInvitationResponse(
    AdminTeamSchema
):
    admin_id: uuid.UUID
    role: AdminRole
    activated_at: datetime


# ============================================================
# ROLE MANAGEMENT
# ============================================================


class ChangeAdminRoleRequest(
    AdminTeamSchema
):
    role: AdminRole


class AdminRoleChangeResponse(
    AdminTeamSchema
):
    admin_id: uuid.UUID
    role: AdminRole
    security_version: int


# ============================================================
# ENABLE / DISABLE
# ============================================================


class AdminAccountStateResponse(
    AdminTeamSchema
):
    admin_id: uuid.UUID
    is_active: bool
    disabled_at: datetime | None
    security_version: int


# ============================================================
# SESSION REVOCATION
# ============================================================


class AdminSessionRevocationResponse(
    AdminTeamSchema
):
    admin_id: uuid.UUID
    security_version: int
    sessions_invalidated: bool = True

File: frontend/src/types/admin-team.ts
export type AdminRole =
  | "reviewer"
  | "admin"
  | "super_admin";


export interface AdminTeamMember {
  id: string;
  email: string;
  display_name: string;
  role: AdminRole;
  is_active: boolean;
  mfa_enabled: boolean;
  last_login_at: string | null;
  invited_at: string | null;
  activated_at: string | null;
  disabled_at: string | null;
  security_updated_at: string | null;
  created_at: string;
  updated_at: string;
}


export interface AdminTeamListResponse {
  items: AdminTeamMember[];
}


export interface AdminInvitation {
  id: string;
  email: string;
  role: AdminRole;
  invited_by_admin_id: string;
  expires_at: string;
  created_at: string;
  updated_at: string;
  accepted_at: string | null;
  accepted_admin_id: string | null;
  revoked_at: string | null;
  revoked_by_admin_id: string | null;
  revoke_reason: string | null;
}


export interface AdminInvitationListResponse {
  items: AdminInvitation[];
}


export interface CreateAdminInvitationPayload {
  email: string;
  role: AdminRole;
  expires_at: string;
}


export interface CreatedAdminInvitationResponse {
  invitation: AdminInvitation;
}


export interface RevokeAdminInvitationPayload {
  reason: string | null;
}


export interface AcceptedAdminInvitationResponse {
  admin_id: string;
  role: AdminRole;
  activated_at: string;
}


export interface ChangeAdminRolePayload {
  role: AdminRole;
}


export interface AdminRoleChangeResponse {
  admin_id: string;
  role: AdminRole;
  security_version: number;
}


export interface AdminAccountStateResponse {
  admin_id: string;
  is_active: boolean;
  disabled_at: string | null;
  security_version: number;
}


export interface AdminSessionRevocationResponse {
  admin_id: string;
  security_version: number;
  sessions_invalidated: boolean;
}
